tex/generic/pgf/utilities/pgfkeys.code.tex

```diff
@@ -74,6 +74,102 @@
 \fi
 
 
+%%% TRACING
+
+%% The tracing code is based on the draft of package "trace-pgfkeys.sty"
+%% (2012/02/06) by Ryan Reich <ryan.reich@gmail.com>.
+
+% If catcode of & is 9 (ignore; most likely set by \tracepgfkeys below), the
+% tracing code, all of which is preceded by this character, will be loaded.
+% Otherwise, we set catcode of & to 14 (comment), and the tracing code will not
+% be loaded.
+\ifnum\catcode`\&=9
```

Comment:
> I appreciate the effort, but the proposed implementation is quite invasive and I'm not sure I want to have this in pgfkeys.

Respectfully, I disagree about the invasiveness.  The fate of the original trace-pgfkeys shows that any non-integrated approach is too hard to maintain, and any integrated implementation will have to add *something*. As far as I see, the proposed implementation adds the bare minimum — tracing code sprinkled into the original macros and clearly distinguishable from non-tracing code (due to the presence of the guard `&`). I would call this minimally invasive.

> Would it maybe be sufficient to only expose tracing hooks inside pgfkeys macros that are then implemented by a separate pgfkeys library?

Well, the hooks approach has the (one and only) advantage of being less intrusive when one is reading `pgfkeys.code.tex`.  But in my opinion, this advantage is outweighed by the disadvantages.

* The added layer of complexity increases the maintenance workload.  In order to add or change some tracing code, one would need to match up hook names in the original and the tracing code (residing in different files).

* It is more error-prone. It would be easy to make a mistake when matching the hook names (above); in contrast, you can't put the tracing code in a wrong place with the hook-less approach. Also, what if whoever proposes a new hook forgets to define it?  At least a dummy would need to be defined.

* It pollutes the string space (many new control sequences).

Finally, the hooks approach still does not solve the issue of the "messy catcode business", raised below.  The hook invocations should still occur behind the guard (`&`), otherwise we are polluting the runtime content of pfgkeys macros — and not doing this, rather than the visual outlook of the code, was my primary concern for the implementation.

> I find the whole catcode business going on here quite messy. The code becomes quite unreadable and the maintenance burden is quite high.

I don't understand how these two comments are related. The catcode logic in the "preamble" is one thing; the readability of the code, i.e. the pgfkeys macros, is another (and apparently a matter of taste).

On the subject of the catcodes. The approach is very similar to the familiar, tried-and-true .dtx approach.  So how is this messy? (Of course, I'm willing to entertain an alternative idea, it's just that I don't have one.)  Let me try to clarify the catcode logic: the bulk of `pgfkeys.code.tex` is read either with `&` having catcode comment (no tracing) or ignore (tracing); the catcode is restored at the end of the file.  (I have chosen `&` because it was not used in `pgfkeys.code.tex` and because it is easily detectable in any editor with TeX syntax highlighting.)


> On top of implementation logic many macros now also have tracing logic. Changing a macro now also requires changing the tracing.

Well, this is sort of the point, isn't it?  (Ok, I don't know if I would call it tracing *logic*, but that's semantics.)  Yes, if we are to have tracing integrated into `pgfkeys`, changing the tracing info upon changing a (core) macro would be required; or at the very least, it would be the polite thing to do.  The real question is then perhaps, do we want to have integrated tracing?  Do we want to have tracing at all?  I have blindly assumed that this is the case, as tracing was immensely useful for me in the past (and I cannot be the only one, if Ryan took the effort of creating `trace-pgfkeys` in the first place), but maybe my assumption was flawed.

It's perhaps also worth pointing out that there shouldn't be much of tracing-related maintenance required.  I mean, how often does the core of pgfkeys change?  I have the impression that any core changes will be very limited, lest they break the ton of existing code using the utility.  I looked through the changes to `pgfkeys.code.tex` in the past five years (the extent shown on GitHub): there was a single change which would require the addition of new tracing code.  (Note that most often, new handlers and such require no tracing code, e.g. the existing tracing code shows quite nicely what's happening with `.evaluate`.)

One final note on maintenance.  In my opinion, it is much easier for whoever develops/changes a macro to update the tracing code.  They know what they are doing and why.  It is an extra minute of work to add tracing.  In contrast, it can take hours for an outsider (like myself) to figure out what the particular piece of code is doing, and trace it properly.  And we can also get it wrong.